docs: haw-sequencer (#3367)
OffchainLabs/arbitrum-docs: docs/how-arbitrum-works/deep-dives/batchposter.mdx

---
title: The Batch Poster
description: 'Learn the fundamentals of the Arbitrum Batch Poster.'
author: pete-vielhaber
sme: Mehdi
user_story: As a current or prospective Arbitrum user, I need to learn more about the batch poster mechanism/protocol.
content_type: concept
---

Batch posting is a fundamental process for the Sequencer's operation in Arbitrum. It involves collecting multiple child chain transactions, organizing them into batches, compressing the data to reduce size, and sending these batches to the Sequencer Inbox contract on the parent chain. This mechanism is crucial for ensuring that transactions are securely recorded on the parent chain blockchain while optimizing for costs and performance.

Understanding batch posting is essential for grasping how Arbitrum achieves scalability and cost-efficiency without compromising security. By delving into these subtopics, you’ll gain insight into the Sequencer’s role in optimizing transaction throughput and minimizing fees, as well as the innovative solutions implemented to address the challenges of parent chain data pricing.

<ImageZoom src="/img/haw-batching.svg" alt="Batching" className="img-600px" />

> #### Batch formation criteria:

Reviewer: ```suggestion
#### Batch formation criteria
```

- **Size thresholds**: Batch formation occurs when accumulated transactions reach a predefined size limit. This limit ensures that the fixed costs of posting data to the parent chain are amortized over more transactions, improving cost efficiency.
- **Time constraints**: The Sequencer also monitors the elapsed time since the last batch update to prevent undue delays. Upon reaching the maximum time threshold, the Sequencer will create a batch with the transactions collected so far, even if the batch doesn’t meet the size threshold.

> #### Batch creation process:

Reviewer: ```suggestion
#### Batch creation process
```

12. **Send it** via the `DataPoster`.
13. **Update the backlog estimate** so future batches compress and bid appropriately.

### DataPoster: L1 transaction management

#### Queue storage backends

1. The DataPoster keeps a queue of L1 transactions it has sent but not yet seen confirmed.
2. That queue needs to survive restarts (so it doesn't lose track of nonces), which is why the default stores it in a local database.
3. If several posters share the job, the queue lives in Redis so they all see the same state.
4. On an Arbitrum parent chain there's no public mempool to worry about, so a no-op store is enough.
5. In-memory storage exists mainly for tests and throwaway setups.

#### Transaction posting flow

1. Confirm the nonce is the one expected, so this transaction slots in correctly.
2. Refresh how much L1 balance the poster has to spend.
3. Work out the fee bids (max fee, tip, and blob fee if applicable) for current conditions.
4. Build the right transaction shape—a normal fee transaction, or a blob-carrying transaction.
5. For blob transactions, do the extra cryptographic work (KZG commitments and proofs) that Ethereum requires.

> 6. Sign it weith the poster's key.

Reviewer: ```suggestion
6. Sign it with the poster's key.
```

7. Record it in the queue (so it can be tracked and replaced) and broadcast it to L1.

#### Fee cap formula

1. Start from a baseline target price the operator considers reasonable.
2. Add a term that grows with the **backlog**: the more batches waiting, the higher the bid—and it grows as the square so pressure ramps up fast.
3. Add another term that grows with how **long** this batch has been waiting, also squared.
4. The result: a batch that's both backed up and old bids very aggressively, while a calm, fresh batch bids modestly near the target price.

#### Replace-by-fee (RBF)

1. After sending, the poster keeps watching transactions that haven't been confirmed.

> 2. When a transaction waiting past its scheduled replacement time, it resends the **same** transaction with a higher fee (replace-by-fee).

Reviewer: ```suggestion
2. When a transaction has been waiting past its scheduled replacement time, it resends the **same** transaction with a higher fee (replace-by-fee).
```

### EIP-4844 blob support

#### Decision logic

1. First, blobs must be turned on in config and supported by the parent chain and ArbOS version—otherwise it's calldata, no question.
2. If blobs are possible, it compares the per-byte cost of each lane right now.
3. Whichever is cheaper for the current L1 market wins.
4. To avoid flip-flopping every block (which wastes the mempool-ordering rules above), once it switches off blobs it waits for a run of non-blob batches before switching back.

#### Blob encoding

1. A blob is a fixed-size container of "field elements," and only 31 of every 32 bytes can hold real data.
2. The batch is RLP-encoded, then poured into those 31-byte slots across the blob.
3. The leftover bits from each slot are gathered up and packed separately so no space is wasted.
4. Each finished blob gets a cryptographic commitment and proofs, which is what lets Ethereum verify the blob without storing it forever.

#### Blob transaction construction

1. The actual blobs ride in a "sidecar" attached to the transaction, not in the main calldata.
2. Alongside the blobs go their commitments and proofs, so validators can check them.

> 3. The transaction body itself only carries lightweight `BlobHashes` (verisoned hashes) that point at the blobs.

Reviewer: ```suggestion
3. The transaction body itself only carries lightweight `BlobHashes` (versioned hashes) that point at the blobs.
```

3. Only when the counter hits zero does it try the offchain backend again.
4. This stops the poster from retrying a flaky backend on every single batch and ping-ponging between the two.

#### AnyTrust aggregator

1. The aggregator sends the batch to all the committee's storage nodes at the same time.
2. Each node that stores it signs a receipt promising to keep the data until a deadline.
3. The aggregator waits until enough nodes have signed (enough that at least one honest node must hold the data).
4. It combines those signatures into one compact certificate.
5. That certificate—not the raw data—is what gets posted to L1, which is why AnyTrust is cheaper than putting everything onchain.

#### DA header bytes

1. The very first byte of every posted batch is a tag describing how to read the rest.
2. A reader checks that byte before anything else to know the format (plain Brotli, a DA certificate, blob hashes, etc.).
3. Some tags are combinations—bits OR'd together—so one byte can say "AnyTrust **and** tree-encoded."
4. This lets a single decoder handle every data-availability format just by branching on the header byte.

### Backlog tracking

#### Estimation

> 1. "Backlog" answers: roughly how mnay batches' worth of messages are waiting but not yet posted?

Reviewer: ```suggestion
1. "Backlog" answers: roughly how many batches' worth of messages are waiting but not yet posted?
```

2. It takes the count of unposted messages and divides by the typical messages-per-batch.
3. That typical figure is a rolling average over the last 20 batches, so it adapts smoothly instead of jumping.
4. If the poster only paused because it bumped into L1 time bounds (not real congestion), it reports zero backlog so it doesn't needlessly degrade compression or overbid.

#### Effects of backlog

1. The same backlog figure feeds three different behaviors.
2. It lowers compression quality so batches go out faster when work piles up.

> 3. It raises the fee bid (squared) so batches comopete harder for L1 inclusion.

Reviewer: ```suggestion
3. It raises the fee bid (squared) so batches compete harder for L1 inclusion.
```

### L1 bounds, delay proofs, and the `SequencerInbox`

#### L1 block bounds

1. The L1 contract will reject a batch whose claimed time/block drifts too far from L1 reality, so the poster stays inside a window.
2. The **upper** edge: don't include messages that reference an L1 time/block further ahead than allowed.
3. The **lower** edge: don't include message so old they're near the bottom of the window, because an L1 reorg could invalidate them.
4. The catch: if a message is **forced** to be near the lower edge (e.g., an old delayed message), strictly enforcing the bound would stall the chain—so past a 1-hour threshold the bound check is bypassed to keep things moving.

#### Reorg resistance margin

`ReorgResistanceMargin` (default 10 minutes) adds a safety buffer.

1. Even inside the legal lower bound, posting right at the edge is risky: a small L1 reorg could push the bound past your batch and invalidate it.
2. So the poster keeps an extra 10-minute buffer above that lower edge.
3. If the batch's first real message sits inside that buffer, it waits rather than posts.
4. Waiting a little for more L1 settling is cheaper than having a posted batch rejected.

#### Delay buffer

> 1. The delay buffer is an onchain mechanism that rewards the power for including delayed (L1-submitted) message promptly.

Reviewer: ```suggestion
1. The delay buffer is an onchain mechanism that rewards the poster for including delayed (L1-submitted) messages promptly.
```

#### Delay proofs

Generated when a batch contains delayed messages **and** the delay buffer is enabled **and** updatable.

1. When a batch pulls in delayed messages and the delay buffer is active, the poster attaches a small proof.
2. The proof carries the accumulator value just before the delayed message plus that message's key fields.
3. On L1, the `SequencerInbox` uses it to update its delay-buffer accounting in the same transaction.
4. Because there's extra data to pass, the poster calls a special `*DelayProof` variant of the contract method instead of the plain one.

### Reading batches back: `SequencerInbox` reader

#### Batch header format

Each batch is serialized with a 40-byte header (5 x uint64).

1. Every batch begins with five numbers packed into 40 bytes, then the data.
2. The first four record the time/block window the batch is valid for (the same bounds the poster computed before posting).
3. The fifth records how many delayed messages have been consumed up to this batch.
4. A reader uses these to re-establish the exact context before decoding the payload that follows.

#### Reading flow

> `LookupBatchsInRange` filters `SequencerBatchDelivered` events from the SequencerInbox contract, parses them into `SequencerInboxBatch` structs with time bounds, accumulator hashes, and data location.

Reviewer: ```suggestion
`LookupBatchesInRange` filters `SequencerBatchDelivered` events from the SequencerInbox contract, parses them into `SequencerInboxBatch` structs with time bounds, accumulator hashes, and data location.
```

1. To learn what batches exist, the reader scans L1 for `SequencerBatchDelivered` events in a block range.
2. Each event becomes a structured record: its time bounds, its accumulator hashes, and where to find its data.
3. With that record, the reader knows enough to go fetch and decode the batch's contents.

### Access lists

The `AccessList` function pre-computes storage slots to reduce gas costs.

1. Ethereum charges less for storage slots if you warn it in advance which ones you'll read.
2. The poster knows exactly which contract slots its batch transaction will touch (in the SequencerInbox, the bridge, and optionally a gas refunder).
3. It lists those slots up front in the transaction's access list to get the discount.
4. On an Arbitrum parent chain this trick doesn't pay off (data is already cheap there), so it's switched off for L3s.

### Error Handling and Retries

#### Batch poster error handling

The `Start()` loop uses `EphemeralErrorHandler` instances to suppress transient errors.

1. Many errors are transient (a race, a temporary L1 hiccup), so the poster doesn't shout about them right away—it stays quiet for a minute or so.

> 2. A known set of "expected" errors gets a longer grace period (debug, then warn, then error) before it botehrs an operator.

Reviewer: ```suggestion
2. A known set of "expected" errors gets a longer grace period (debug, then warn, then error) before it bothers an operator.
```
